tests-functional/clients/signals.py

```diff
@@ -3,32 +3,46 @@
 import time
 
 import websocket
+import os
+from pathlib import Path
+from constants import SIGNALS_DIR, LOG_SIGNALS_TO_FILE
+from datetime import datetime
 
 
 class SignalClient:
-
     def __init__(self, ws_url, await_signals):
         self.url = f"{ws_url}/signals"
 
         self.await_signals = await_signals
         self.received_signals = {
             signal: [] for signal in self.await_signals
         }
+        if LOG_SIGNALS_TO_FILE:
+            self.signal_file_path = os.path.join(SIGNALS_DIR, f"signal_{ws_url.split(':')[-1]}_{datetime.now().strftime('%H%M%S')}.log")
+            Path(SIGNALS_DIR).mkdir(parents=True, exist_ok=True)
 
     def on_message(self, ws, signal):
-        signal = json.loads(signal)
-        if signal.get("type") in self.await_signals:
-            self.received_signals[signal["type"]].append(signal)
+        signal_data = json.loads(signal)
+        if LOG_SIGNALS_TO_FILE:
+            self.write_signal_to_file(signal_data)
+
+        signal_type = signal_data.get("type")
+        if signal_type in self.await_signals:
+            self.received_signals[signal_type].append(signal_data)
 
-    def wait_for_signal(self, signal_type, timeout=20):
+    def wait_for_signal(self, signal_type, timeout=20, event_contains=None):
         start_time = time.time()
-        while not self.received_signals.get(signal_type):
+        while True:
+            if self.received_signals.get(signal_type):
+                for event in self.received_signals[signal_type]:
+                    if event_contains is None or event_contains in str(event):
+                        logging.info(
+                            f"Signal {signal_type} containing {event_contains} is received in {round(time.time() - start_time)} seconds")
+                        return event
```

Comment:
We're a bit of fans of early return pattern here 🙏 
```suggestion
            if not self.received_signals.get(signal_type):
                continue
            for event in self.received_signals[signal_type]:
                if event_pattern is not None and event_pattern not in str(event):
                    continue
                logging.info(
                    f"Signal {signal_type} containing {event_contains} is received in {round(time.time() - start_time)} seconds")
                return event
```